eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxWits.hs

```diff
@@ -579,55 +580,93 @@ instance
           )
       txWitnessField 1 =
-        fieldAA
-          addScriptsTxWitsRaw
-          (D nativeScriptsDecoder)
+        fieldAA addScriptsTxWitsRaw (D scriptsDecoder)
       txWitnessField 2 =
         fieldAA
           (\x wits -> wits {atwrBootAddrTxWits = x})
           ( D $
               ifDecoderVersionAtLeast
                 (natVersion @9)
-                ( allowTag setTag
-                    >> mapTraverseableDecoderA (decodeNonEmptyList decCBOR) (Set.fromList . NE.toList)
-                )
+                setDecoder
                 (mapTraverseableDecoderA (decodeList decCBOR) Set.fromList)
           )
       txWitnessField 3 = fieldA addScriptsTxWitsRaw (decodeAlonzoPlutusScript SPlutusV1)
       txWitnessField 4 =
         fieldAA
           (\x wits -> wits {atwrDatsTxWits = x})
-          From
+          ( D $
+              ifDecoderVersionAtLeast
+                (natVersion @12)
+                noDuplicatesDatsDecoder
+                decCBOR
+          )
       txWitnessField 5 = fieldAA (\x wits -> wits {atwrRdmrsTxWits = x}) From
       txWitnessField 6 = fieldA addScriptsTxWitsRaw (decodeAlonzoPlutusScript SPlutusV2)
       txWitnessField 7 = fieldA addScriptsTxWitsRaw (decodeAlonzoPlutusScript SPlutusV3)
       txWitnessField n = invalidField n
       {-# INLINE txWitnessField #-}
 
-      nativeScriptsDecoder :: Decoder s (Annotator (Map ScriptHash (Script era)))
-      nativeScriptsDecoder =
-        ifDecoderVersionAtLeast
-          (natVersion @9)
-          ( allowTag setTag
-              >> mapTraverseableDecoderA (decodeNonEmptyList pairDecoder) (Map.fromList . NE.toList)
-          )
-          (mapTraverseableDecoderA (decodeList pairDecoder) Map.fromList)
+      pairDecoder :: Decoder s (Annotator (ScriptHash, Script era))
+      pairDecoder = fmap (asHashedScriptPair @era . fromNativeScript) <$> decCBOR
+      {-# INLINE pairDecoder #-}
+
+      noDuplicatesDatsDecoder :: Decoder s (Annotator (TxDats era))
+      noDuplicatesDatsDecoder = do
+        allowTag setTag
+        dats <- decodeList decCBOR
+        pure $ TxDats <$> go Map.empty dats
         where
-          pairDecoder :: Decoder s (Annotator (ScriptHash, Script era))
-          pairDecoder = fmap (asHashedScriptPair . fromNativeScript) <$> decCBOR
+          go m [] = pure m
+          go m (x : xs) = do
+            x' <- x
+            let dh = hashData x'
+            if dh `Map.member` m
+              then fail $ "Duplicate dats found: " <> show dh
+              else go (Map.insert dh x' m) xs
+      {-# INLINE noDuplicatesDatsDecoder #-}
+
+      noDuplicatesScriptsDecoder :: Decoder s (Annotator (Map ScriptHash (Script era)))
+      noDuplicatesScriptsDecoder = do
+        allowTag setTag
+        scripts <- decodeList $ fmap (fromNativeScript @era) <$> decCBOR
+        pure $ go Map.empty scripts
+        where
+          go m [] = pure m
+          go m (x : xs) = do
```

Comment:
We need to make sure we avoid space leaks and have decoders as performant as possible
```suggestion
          go !m (x : xs) = do
```